opencog/cogserver/modules/events/AtomSpacePublisherModule.h

```diff
@@ -81,91 +81,91 @@ class AtomSpacePublisherModule;
-        void AVChangedSignal(const Handle& h,
-                             const AttentionValuePtr& av_old,
-                             const AttentionValuePtr& av_new);
-        void TVChangedSignal(const Handle& h,
-                             const TruthValuePtr& tv_old,
-                             const TruthValuePtr& tv_new);
-        void addAFSignal(const Handle& h,
-                         const AttentionValuePtr& av_old,
-                         const AttentionValuePtr& av_new);
-        void removeAFSignal(const Handle& h,
-                         const AttentionValuePtr& av_old,
-                         const AttentionValuePtr& av_new);
+	private:
+		AtomSpace* as;
+		AttentionBank* _attention_bank;
+
+		AtomPtrSignal* _remove_atom_signal;
+		int _remove_atom_connection;
+
+		AtomSignal* _add_atom_signal;
+		int _add_atom_connection;
+
+		TVCHSigl* _tvchange_signal;
+		int _tvchange_connection;
+
+		AVCHSigl* _avchange_signal;
+		int _avchange_connection;
+
+		AVCHSigl* _add_af_signal;
+		int _add_af_connection;
+
+		AVCHSigl* _remove_af_signal;
+		int _remove_af_connection;
+
+		void enableSignals();
+		void disableSignals();
+
+		// TBB
+		tbb::concurrent_bounded_queue<message_t> queue;
+
+		// ZeroMQ
+		zmq::context_t * context;
+		void InitZeroMQ();
+		void proxy();
+
+		void sendMessage(std::string messageType, std::string payload);
+		std::string atomMessage(Json::Value jsonAtom);
+		std::string avMessage(Json::Value jsonAtom, Json::Value jsonAVOld, Json::Value jsonAVNew);
+		std::string tvMessage(Json::Value jsonAtom, Json::Value jsonTVOld, Json::Value jsonTVNew);
+		Json::Value atomToJSON(Handle h);
+		Json::Value tvToJSON(TruthValuePtr tv);
+		Json::Value avToJSON(AttentionValuePtr av);
+		// TODO: add protoatom to JSON functionality
+
+		DECLARE_CMD_REQUEST(AtomSpacePublisherModule, "publisher-enable-signals",
+			 do_publisherEnableSignals,
+			 "Enable AtomSpace event publishing",
+			 "Usage: publisher-enable-signals",
+			 false, false)
+
+		DECLARE_CMD_REQUEST(AtomSpacePublisherModule, "publisher-disable-signals",
+			 do_publisherDisableSignals,
+			 "Disable AtomSpace event publishing",
+			 "Usage: publisher-disable-signals",
+			 false, false)
+
+	public:
+		AtomSpacePublisherModule(CogServer&);
+		virtual ~AtomSpacePublisherModule();
+		virtual void run();
+
+		static const char *id(void);
+		virtual void init(void);
+
+		void atomAddSignal(Handle h);
+		void atomRemoveSignal(AtomPtr atom);
+		void AVChangedSignal(const Handle& h,
+							 const AttentionValuePtr& av_old,
+							 const AttentionValuePtr& av_new);
+		void TVChangedSignal(const Handle& h,
+							 const TruthValuePtr& tv_old,
+							 const TruthValuePtr& tv_new);
+		void addAFSignal(const Handle& h,
+						 const AttentionValuePtr& av_old,
+						 const AttentionValuePtr& av_new);
+		void removeAFSignal(const Handle& h,
+						 const AttentionValuePtr& av_old,
+						 const AttentionValuePtr& av_new);
```

Comment:
Oh, and also, it would be great if this could be designed in some way to not use signals, especially the atomspace add/remove signals. The problem is that these signals run synchronously, in the thread adding/removing the atom.  This slows down performance, which is already poor. I don't yet know what a good substitute would be, but its something to think about, in the back of one's mind.

Reply:
Replacing the signals seems like it would be a whole lot of change. 
But since the signals are basically callbacks to a user-defined function, the performance of the user-defined function ultimately affects the performance of the atomspace (while waiting for the user function to finish execution) 
it should be something that is called and the caller can just go on with the rest of its activities. 
filling a queue like zmq does.